// runescape-api/index.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const osrs = __importStar(require("./osrs"));
const bestiary = __importStar(require("./runescape/bestiary"));
exports.bestiary = bestiary;
const clan = __importStar(require("./runescape/clan"));
exports.clan = clan;
const grandexchange = __importStar(require("./runescape/grandexchange"));
exports.grandexchange = grandexchange;
const hiscores = __importStar(require("./runescape/hiscores"));
exports.hiscores = hiscores;
const miscellaneous = __importStar(require("./runescape/miscellaneous"));
exports.miscellaneous = miscellaneous;
const runemetrics = __importStar(require("./runescape/runemetrics"));
exports.runemetrics = runemetrics;
exports.default = {
    bestiary,
    clan,
    grandexchange,
    hiscores,
    miscellaneous,
    runemetrics,
    osrs,
};
